### PR TITLE
Parpack

### PR DESCRIPTION
**Pull request by davydden**
Parpack solver with a step-36 example.

### CODE REVIEW

**Comment by tamiko** [comment]
So time to look at the pull request again. Can you please rebase to current master?

**Comment by tamiko** [comment]
OK. There is one cosmetic issue. With the change of <code>config.h.in</code> the <code>#cmakedefine ARPACK_WITH_PARPACK</code> is now out of place. Please apply
```
diff --git a/include/deal.II/base/config.h.in b/include/deal.II/base/config.h.in
index 6909946..2644bd7 100644
--- a/include/deal.II/base/config.h.in
+++ b/include/deal.II/base/config.h.in
@@ -62,6 +62,8 @@
 #cmakedefine DEAL_II_WITH_UMFPACK
 #cmakedefine DEAL_II_WITH_ZLIB
 
+#cmakedefine ARPACK_WITH_PARPACK
+
 
 /***********************************************************************
  * Compiler bugs:
@@ -189,7 +191,6 @@
 /*
  * deal.II:
  */
-#cmakedefine ARPACK_WITH_PARPACK
 
 #define DEAL_II_VERSION_GTE(major,minor,subminor) \
  ((DEAL_II_VERSION_MAJOR * 10000 + \
```

**Comment by bangerth** [comment]
Shouldn't this also be `DEAL_II_ARPACK_WITH_PARPACK` instead?

File: include/deal.II/lac/parpack_solver.h [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+  // other resources:
+  //    http://acts.nersc.gov/superlu/example5/pnslac.c.html
+  //    https://github.com/phpisciuneri/tijo/blob/master/dvr_parpack.cpp
+
+}
+
+/**
+ * Interface for using PARPACK. PARPACK is a collection of Fortran77 subroutines
+ * designed to solve large scale eigenvalue problems.  Here we interface to
+ * the routines <code>pdneupd</code>, <code>pdseupd</code>, <code>pdnaupd</code>, <code>pdsaupd</code> of PARPACK.
```
Comment (tamiko) [inline include/deal.II/lac/parpack_solver.h]: Please break all overlong lines in comments at at most 80 characters.

File: include/deal.II/lac/parpack_solver.h [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+  DeclException2 (ExcInvalidNumberofArnoldiVectors, int, int,
+                  << "Number of Arnoldi vectors " << arg1
+                  << " is larger that the size of the matrix " << arg2);
+
+  DeclException2 (ExcSmallNumberofArnoldiVectors, int, int,
+                  << "Number of Arnoldi vectors " << arg1
+                  << " is too small to obtain " << arg2
+                  << " eigenvalues");
+
+  DeclException1 (ExcArpackIdo, int, << "This ido " << arg1
```
Comment (tamiko) [inline include/deal.II/lac/parpack_solver.h]: Can you rename the exceptions to <code>PArpack</code>?

**Comment by tamiko** [comment]
@bangerth: Yes, this is a bit of a problem. Unfortunately, <code>#cmakedefine</code> is not able to expand <code>DEAL_II_FOO</code> depending on a variable called <code>FOO</code>. It has to be the same name.

The only solution for this that I see is to refactor parpack into its own feature that sets up <code>DEAL_II_WITH_PARPACK</code>.

File: include/deal.II/lac/parpack_solver.h (hunk 1 above)
Comment (davydden) [inline include/deal.II/lac/parpack_solver.h]: any way to do automatically (in Eclipse)?

Comment (bangerth) [inline include/deal.II/lac/parpack_solver.h]: Not that I know of :-(

Comment (bangerth) [inline include/deal.II/lac/parpack_solver.h]: In emacs "Esc-q" does it.

Comment (davydden) [inline include/deal.II/lac/parpack_solver.h]: i don't use e-macs... eclipse does not seem to format comments properly

**Comment by bangerth** [comment]
I don't think I understand. There is a place in the cmake scripts where we set
```
IF(NOT PARPACK_LIBRARY MATCHES "-NOTFOUND")
  SET(ARPACK_WITH_PARPACK TRUE)
ELSE()
  SET(ARPACK_WITH_PARPACK FALSE)
ENDIF
```
Why can't we rename the variable `ARPACK_WITH_PARPACK` to `DEAL_II_ARPACK_WITH_PARPACK` and do the corresponding change in `config.h`

**Comment by tamiko** [comment]
> Why can't we rename the variable `ARPACK_WITH_PARPACK` to
> `DEAL_II_ARPACK_WITH_PARPACK` and do the corresponding change in
> `config.h`

Because by convention <code>FindFOO</code> only exports
<code>FOO_***</code>. And I hate to break this rule even more than I
hate to export something not prepended by <code>DEAL_II_***</code> in
<code>config.h</code>.

**Comment by bangerth** [comment]
So what if you call the variable `DEAL_II_WITH_PARPACK`? You're setting the variable in both arms of the `IF`, so it's not like any of the values that cmake comes up with is used anyway. And we have similar `DEAL_II_WITH_*` for our other packages as well.

**Comment by tamiko** [comment]
*ugh* No. <code>DEAL_II_WITH_*</code> exclusively reserved for "features".

@davydden please do:
 - a <code>SET(DEAL_II_ARPACK_WITH_PARPACK ARPACK_WITH_PARPACK)</code> in <code>configure_arpack.cmake</code> after the call to <code>CONFIGURE_FEATURE(ARPACK)</code>
 - <code>#cmakedefine ARPACK_WITH_PARPACK</code> to <code>#cmakedefine DEAL_II_ARPACK_WITH_PARPACK</code> IN <code>config.h.in</code>.

**Comment by davydden** [comment]
Let me know when it is alright to stash.

**Comment by tamiko** [comment]
Please rebase and squash a last time.

In my opinion this pull-request can be merged. I dislike the amount of code duplication between <code>arpack_solver.h</code> and <code>parpack_solver</code>, but I see no immediate way to avoid it in a straightforward/nice manner.

Can somebody else please have a second look and merge? (After rebase and squash).

**Comment by davydden** [comment]
before squashing, does anyone know how to suppress output of
```
DEAL:cg::Starting value 0.357308
DEAL:cg::Convergence step 55 value 0
...
```
in `step-36_parpack_Trilinos.cc`? What I check is resulting eigenvalues and not the iterations of `IterativeInverse` class.

**Comment by tamiko** [comment]
Am 19. Jan 2015, 21:54 schrieb Denis Davydov <notifications@github.com>:

> before squashing, does anyone know how to suppress output of
> ```
> DEAL:cg::Starting value 0.357308
> DEAL:cg::Convergence step 55 value 0
> ...
> ```
> in `step-36_parpack_Trilinos.cc`? What I check is resulting
> eigenvalues and not the iterations of `IterativeInverse` class.

If the output is completely unimportant, you can simply do something like

  const unsigned int previous_depth = deallog.depth_file(0);
  [...]
  deallog.depth_file(previous_depth);

If not, have a look at the macro
  
  check_solver_within_range(SOLVER_COMMAND, CONTROL_COMMAND, MIN_ALLOWED, MAX_ALLOWED)

defined in tests.h

**Comment by davydden** [comment]
thanks, @tamiko . Will do that, rebase and squash...

**Comment by davydden** [comment]
I messed up something in the last couple of commits... :( all tests fail, need to have a closer look...

**Comment by davydden** [comment]
can someone pls try running `arpack` tests? I have segmentation errors...

**Comment by davydden** [comment]
p.s. it seems there is an issue with make setup_tests and cmake 3.1.0:
```
CMake Warning (dev) at /Users/davydden/libs-sources/deal.ii/davydden/tests/setup_testsubproject.cmake:70 (IF):
  Policy CMP0054 is not set: Only interpret if() arguments as variables or
  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  Quoted variables like "DEAL_II_BINARY_DIR" will no longer be dereferenced
  when the policy is set to NEW.  Since the policy is not set the OLD
  behavior will be used.
Call Stack (most recent call first):
  CMakeLists.txt:2 (INCLUDE)
This warning is for project developers.  Use -Wno-dev to suppress it.
```
and then everything freezes...

**Comment by tamiko** [comment]
Am 20. Jan 2015, 00:45 schrieb Denis Davydov <notifications@github.com>:

> p.s. it seems there is an issue with make setup_tests and cmake
> 3.1.0:

The warning is now fixed.

> and then everything freezes...

I cannot reproduce any freezes here - maybe it just takes a long time?

**Comment by davydden** [comment]
Can you please try running `arpack` tests on this branch? I suddenly have issues here... not sure if this is my toolchain or something changed in deal.ii that exposed some bugs in the code

**Comment by tamiko** [comment]
Am 20. Jan 2015, 12:09 schrieb Denis Davydov <notifications@github.com>:

> Can you please try running `arpack` tests on this branch? I suddenly
> have issues here... not sure if this is my toolchain or something
> changed in deal.ii that exposed some bugs in the code

Well, I cannot test due to missing metis:

4: An error occurred in line <70> of file </var/tmp/portage/sci-libs/dealii-9999/work/dealii-9999/source/lac/sparsity_tools.cc> in function
4:     void dealii::SparsityTools::partition(const dealii::SparsityPattern &, const unsigned int, std::vector<unsigned int> &)
4: The violated condition was: 
4:     false
4: The name and call sequence of the exception was:
4:     ExcMETISNotInstalled()
4: Additional Information: 

If this is intended, please constrain the test with ".with_metis=true."
and note that I will not be able to test it on simserv at all.

Best,
Matthias

**Comment by davydden** [comment]
hm... that's probably another reason why it's good to have p4est-alike method applicable to shared triangulation as discussed in this PR https://github.com/dealii/dealii/pull/212

What I could do to make it running is, supposedly, get rid of Metis partitioning and hand-partition DoFs for each processor sequentially. Assembly would, of course, need to go over all cells on each processor. Does this sound good?

**Comment by davydden** [comment]
Better yet, I could probably remove calling `metis` and just assign cells to each processor myself and keep the rest as it is?

**Comment by tamiko** [comment]
> Better yet, I could probably remove calling metis and just assign
> cells to each processor myself and keep the rest as it is?

Sounds good. And it has the additional advantage of being not dependend
on the actual partition metis delivers.

**Comment by tamiko** [comment]
Looks good. Please rebase a last time and squash.

**Comment by davydden** [comment]
Now I will try to make tests independent on `metis`...

**Comment by davydden** [comment]
@tamiko could you please try testing again? should not required `metis` now.

On my macbook I have:
```
* thread #1: tid = 0x228de1, 0x000000011321106f libscalapack.dylib`dgamx2d_ + 91, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x000000011321106f libscalapack.dylib`dgamx2d_ + 91
libscalapack.dylib`dgamx2d_ + 91:
-> 0x11321106f:  movq   (%rax,%rdi,8), %r14
   0x113211073:  jae    0x113211079               ; dgamx2d_ + 101
   0x113211075:  orb    $0x20, %r12b
   0x113211079:  movb   (%rsi), %dl
```
which is likely to be related to my installation of `scalapack` and, hopefully, not on the code in PR.

**Comment by tamiko** [comment]
Thanks for the work on it.

**Comment by tamiko** [comment]
*ugh* We error out badly. [1,2]. Can you please have a look? The latter can be resolved by prepending <code>Triangulation</code> and <code>Point</code> with <code>dealii::</code>. 

[1] http://cdash.kyomu.43-1.org/testDetails.php?test=5685911&build=3180
[2] http://cdash.kyomu.43-1.org/testDetails.php?test=5688267&build=3180

**Comment by davydden** [comment]
@tamiko can you please un-merge it? I pushed some updated but they are not shown here...

**Comment by tamiko** [comment]
@davydden reverted.

You can find the fixups on top of your current parpack branch at [1], pull-request at [2]

[1] https://github.com/tamiko/dealii/tree/parpack
[2] https://github.com/davydden/dealii/pull/1

**Comment by davydden** [comment]
it seems i will need to create a new PR. Will do that when i make things working locally

**Comment by bangerth** [comment]
The test didn't work (and required prefixing everything with `deal::`) because it doesn't `#include "../tests.h"` as is necessary for all tests. Please add that when you fix them up.

**Comment by davydden** [comment]
@bangerth  unfortunately, that's not the main reason.

Late november tests were running fine for me locally. So it's either i screw up something while cleaning up the PR, or something changed. I need to give it a closer look.

**Comment by bangerth** [comment]
Yes, these are separate problems. The segfault is certainly not caused by the lack of the include file, but the compiler errors are. In any case, both have to be addressed :-)

**Comment by davydden** [comment]
maybe there is indeed something wrong with current parpack:

`Invalid`:

```
==84000== Invalid read of size 8
==84000==    at 0x113B4906F: dgamx2d_ (in /usr/local/Cellar/scalapack/2.0.2_1/lib/libscalapack.dylib)
==84000==    by 0x113D121B3: pdlamch_ (in /usr/local/Cellar/scalapack/2.0.2_1/lib/libscalapack.dylib)
==84000==    by 0x1149100C3: pdsaup2_ (in /usr/local/Cellar/arpack/3.1.4_1/libexec/lib/libparpack.2.dylib)
==84000==    by 0x1288012CF: ???
==84000==    by 0x128801A4F: ???
==84000==    by 0x128801A4F: ???
==84000==    by 0x1287FF06F: ???
==84000==    by 0x12880128F: ???
==84000==    by 0x12880128F: ???
==84000==    by 0xC0000016B: ???
==84000==    by 0x7FFF0000016B: ???
==84000==    by 0x1287F5C0F: ???
==84000==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
```

and  `definitely lost`:

```
==84000== 105 bytes in 1 blocks are definitely lost in loss record 9,637 of 11,687
==84000==    at 0x1000CD5CB: malloc (in /usr/local/Cellar/valgrind/HEAD/lib/valgrind/vgpreload_memcheck-amd64-darwin.so)
==84000==    by 0x119104F45: opal_os_path (in /usr/local/Cellar/open-mpi/1.8.4/lib/libopen-pal.6.dylib)
==84000==    by 0x119088652: orte_ess_base_app_abort (in /usr/local/Cellar/open-mpi/1.8.4/lib/libopen-rte.7.dylib)
==84000==    by 0x1190871E2: orte_errmgr_base_abort (in /usr/local/Cellar/open-mpi/1.8.4/lib/libopen-rte.7.dylib)
==84000==    by 0x11F4ACBBE: abort_peers (in /usr/local/Cellar/open-mpi/1.8.4/lib/openmpi/mca_errmgr_default_app.so)
==84000==    by 0x117A3F567: ompi_mpi_abort (in /usr/local/Cellar/open-mpi/1.8.4/lib/libmpi.1.dylib)
==84000==    by 0x117C8AB84: PetscSignalHandlerDefault (in /usr/local/Cellar/petsc/3.5.2_2/real/lib/libpetsc.3.5.2.dylib)
==84000==    by 0x117C8AEC2: PetscSignalHandler_Private (in /usr/local/Cellar/petsc/3.5.2_2/real/lib/libpetsc.3.5.2.dylib)
==84000==    by 0x23805BC0A: ???
==84000==    by 0x113D121B3: pdlamch_ (in /usr/local/Cellar/scalapack/2.0.2_1/lib/libscalapack.dylib)
==84000==    by 0x1149100C3: pdsaup2_ (in /usr/local/Cellar/arpack/3.1.4_1/libexec/lib/libparpack.2.dylib)
==84000==    by 0x1288012CF: ???
```

**Comment by davydden** [comment]
For a record, i tried running with `lldb` debugger:

```
(lldb) r
Process 87381 launched: '/Users/davydden/libs-sources/deal.ii/builds/parpack/tests/arpack/step-36_parpack.debug/step-36_parpack.debug' (x86_64)
Process 87381 stopped
* thread #1: tid = 0x20a531, 0x000000011321d06f libscalapack.dylib`dgamx2d_ + 91, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x000000011321d06f libscalapack.dylib`dgamx2d_ + 91
libscalapack.dylib`dgamx2d_ + 91:
-> 0x11321d06f:  movq   (%rax,%rdi,8), %r14
   0x11321d073:  jae    0x11321d079               ; dgamx2d_ + 101
   0x11321d075:  orb    $0x20, %r12b
   0x11321d079:  movb   (%rsi), %dl
```
which points to `dgamx2d_` function again.

**Comment by davydden** [comment]
guess what? when the HEAD version of `scalapack` is build, all tests run fine and no error appear (run on the current version of the branch where the only changes are adding `test.h` to the test file) :

```
$ ctest -R "arpack"
Test project /Users/davydden/libs-sources/deal.ii/builds/parpack
    Start 1: arpack/step-36_arpack.debug
1/6 Test #1: arpack/step-36_arpack.debug ........................   Passed    0.89 sec
    Start 2: arpack/step-36_arpack.release
2/6 Test #2: arpack/step-36_arpack.release ......................   Passed    0.84 sec
    Start 3: arpack/step-36_parpack.mpirun=3.debug
3/6 Test #3: arpack/step-36_parpack.mpirun=3.debug ..............   Passed    1.05 sec
    Start 4: arpack/step-36_parpack.mpirun=3.release
4/6 Test #4: arpack/step-36_parpack.mpirun=3.release ............   Passed    0.91 sec
    Start 5: arpack/step-36_parpack_trilinos.mpirun=3.debug
5/6 Test #5: arpack/step-36_parpack_trilinos.mpirun=3.debug .....   Passed    1.27 sec
    Start 6: arpack/step-36_parpack_trilinos.mpirun=3.release
6/6 Test #6: arpack/step-36_parpack_trilinos.mpirun=3.release ...   Passed    1.03 sec

100% tests passed, 0 tests failed out of 6

Total Test time (real) =   6.00 sec
```

so technically speaking the PR was ok, the `Scalapack` had a bug...
What do we do in that case?

**Comment by davydden** [comment]
@tamiko @bangerth what shall we do in this case?

**Comment by davydden** [comment]
p.s. I created a post on Scalapack forum http://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=13&t=4676

**Pull request merged**